ui/vty/src/Ariadne/UI/Vty/Widget/Wallet.hs

```diff
@@ -163,59 +187,59 @@ handleWalletWidgetEvent langFace@UiLangFace{..} = \case
-      BrickWalletSendPass -> do
-        field <- use walletFieldSendPassL
-        get >>= lift . handleFormFieldEvent BrickWalletSendPass (B.VtyEvent vtyEv) walletFieldSendPassL field >>= put
-      _ ->
-        return ()
-  WalletBalanceCommandResult commandId result -> do
+  UiCommandResult commandId (UiBalanceCommandResult result) -> do
     walletBalanceL %= \case
       BalanceResultWaiting commandId' | commandId == commandId' ->
         case result of
           UiBalanceCommandSuccess balance -> BalanceResultSuccess balance
           UiBalanceCommandFailure err -> BalanceResultError err
       other -> other
-  WalletSendCommandResult commandId result -> do
+  UiCommandResult commandId (UiSendCommandResult result) -> do
     walletSendResultL %= \case
       SendResultWaiting commandId' | commandId == commandId' ->
         case result of
           UiSendCommandSuccess tr -> SendResultSuccess tr
           UiSendCommandFailure err -> SendResultError err
       other -> other
+  _ ->
+    return ()
 
 ----------------------------------------------------------------------------
 -- Actions
 ----------------------------------------------------------------------------
 
-performSendTransaction
-  :: UiLangFace
-  -> StateT WalletWidgetState (B.EventM BrickName) ()
-performSendTransaction UiLangFace{..} = do
-  address <- use walletSendAddressL
-  amount <- use walletSendAmountL
+addOutput :: Monad m => StateT WalletWidgetState (StateT (WidgetInfo WalletWidgetState p) m) ()
+addOutput = do
+  idx <- walletSendNextOutputL <<+= 1
+  walletSendOutputsL %= Map.insert idx (WalletSendOutput "" "")
+
+  lift $ do
+    addWidgetChild (WidgetNameWalletSendAddress idx) $
+      initEditWidget $ widgetParentLens $ walletSendOutputsL . at idx . unsafeFromJust . walletSendAddressL
+    addWidgetChild (WidgetNameWalletSendAmount idx) $
+      initEditWidget $ widgetParentLens $ walletSendOutputsL . at idx . unsafeFromJust . walletSendAmountL
+    addWidgetChild (WidgetNameWalletSendRemove idx) $
+      initButtonWidget "-"
+    addWidgetEventHandler (WidgetNameWalletSendRemove idx) $ \case
+      WidgetEventButtonPressed -> removeOutput idx
+      _ -> return ()
+
+removeOutput :: Int -> WidgetEventM WalletWidgetState p ()
+removeOutput idx = do
+  walletSendOutputsL %= Map.delete idx
+
+performSendTransaction :: WidgetEventM WalletWidgetState p ()
+performSendTransaction = do
+  UiLangFace{..} <- use walletLangFaceL
+  inputs <- fmap (\WalletSendOutput{..} -> (walletSendAddress, walletSendAmount)) <$> uses walletSendOutputsL Map.elems
   passphrase <- use walletSendPassL
   use walletSendResultL >>= \case
     SendResultWaiting _ -> return ()
-    _ -> liftIO (langPutUiCommand $ UiSend address amount passphrase) >>=
+    _ -> liftIO (langPutUiCommand $ UiSend inputs passphrase) >>=
       assign walletSendResultL . either SendResultError SendResultWaiting
+
+unsafeFromJust :: Lens' (Maybe a) a
+unsafeFromJust = lens fromJust setJust
```

Comment:
I regret spending time on that, since I couldn't improve this code.
One possible way is to use `Traversal'` instead of `Lens'`, as it allows for zero or more values, then you could use `. _Just .` unstead of `unsafeFromJust`. But that probably isn't worth it.